In [1]:
# Importação de bibliotecas
import pandas as pd
import numpy as np

In [4]:
# Carga das bases de dados
order_payments = pd.read_csv('../base de dados/olist_order_payments_dataset.csv')
products = pd.read_csv('../base de dados/olist_products_dataset.csv')

### 1. Análise exploratória inicial

#### 1.1 Base de dados de pagamentos

In [9]:
order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [10]:
order_payments.tail()

order_id  payment_sequential payment_type  \
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58

In [5]:
order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


A base de dados é constituída por 5 características e 103.886 registros e não há dados faltantes em qualquer característica.

In [11]:
order_payments.describe()

payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

Uma compra pode ter vários pagamentos, os quais são identificados por um número sequencial (payment_sequantial) iniciado em 1.

Uma compra também pode ser parcelada, a coluna "payment_installments" indica a quantidade de parcelas.

Os tipos de pagamentos são os seguintes:

In [52]:
order_payments['payment_type'].value_counts()


credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: payment_type, dtype: int64

Os pagamentos são feitos majoritariamente por cartão de crédito, o que representa quase 74% do total.

In [53]:
order_payments['payment_type'].value_counts(True)

credit_card    0.739224
boleto         0.190440
voucher        0.055590
debit_card     0.014718
not_defined    0.000029
Name: payment_type, dtype: float64

A quantidade total de compras (99.440) é menor que a quantidade de registros na base de cálculo (103.886), o que demonstra que uma compra pode ter vários pagamentos, conforme mencionamos acima.

In [35]:
order_payments['order_id'].unique().size

99440

In [26]:
order_payments['order_id'].value_counts()

fa65dad1b0e818e3ccc5cb0e39231352    29
ccf804e764ed5650cd8759557269dc13    26
285c2e15bebd4ac83635ccc563dc71f4    22
895ab968e7bb0d5659d16cd74cd1650c    21
fedcd9f7ccdc8cba3a18defedd1a5547    19
                                    ..
6d2a30c9b7dcee3ed507dc9a601f99e7     1
a7737f6d9208dd56ea498a322ed3c37f     1
646e62df54f3e236eb6d5ff3b31429b8     1
e115da7a49ec2acf622e1f31da65cfb9     1
28bbae6599b09d39ca406b747b6632b1     1
Name: order_id, Length: 99440, dtype: int64

Por exemplo, a compra com 26 pagamentos, identificada pelo id ccf804e764ed5650cd8759557269dc13, teve 1 pagamento com cartão de crédito e os demais com voucher.

In [57]:
order_payments[order_payments['order_id'] == 'ccf804e764ed5650cd8759557269dc13']

order_id  payment_sequential payment_type  \
4346   ccf804e764ed5650cd8759557269dc13                  19      voucher   
14362  ccf804e764ed5650cd8759557269dc13                   1  credit_card   
16279  ccf804e764ed5650cd8759557269dc13                  14      voucher   
18876  ccf804e764ed5650cd8759557269dc13                  15      voucher   
21447  ccf804e764ed5650cd8759557269dc13                   6      voucher   
32393  ccf804e764ed5650cd8759557269dc13                  26      voucher   
38001  ccf804e764ed5650cd8759557269dc13                   2      voucher   
39132  ccf804e764ed5650cd8759557269dc13                  25      voucher   
40369  ccf804e764ed5650cd8759557269dc13                   5      voucher   
40387  ccf804e764ed5650cd8759557269dc13                   3      voucher   
41414  ccf804e764ed5650cd8759557269dc13                   7      voucher   
45857  ccf804e764ed5650cd8759557269dc13                  11      voucher   
49137  ccf804e764ed5650cd8759557269dc13                  13      voucher   
51816  ccf804e764ed5650cd8759557269dc13                  24      voucher   
53622  ccf804e764ed5650cd8759557269dc13                  22      voucher   
60241  ccf804e764ed5650cd8759557269dc13                  23      voucher   
71309  ccf804e764ed5650cd8759557269dc13                  10      voucher   
76655  ccf804e764ed5650cd8759557269dc13                  20      voucher   
79324  ccf804e764ed5650cd8759557269dc13                   8      voucher   
87981  ccf804e764ed5650cd8759557269dc13                   9      voucher   
88364  ccf804e764ed5650cd8759557269dc13                  18      voucher   
91510  ccf804e764ed5650cd8759557269dc13                  12      voucher   
93531  ccf804e764ed5650cd8759557269dc13                  21      voucher   
94873  ccf804e764ed5650cd8759557269dc13                  16      voucher   
95153  ccf804e764ed5650cd8759557269dc13                   4      voucher   
96566  ccf804e764ed5650cd8759557269dc13                  17      voucher   

       payment_installments  payment_value  
4346                      1           1.68  
14362                     1           2.20  
16279                     1           0.97  
18876                     1           1.11  
21447                     1           2.26  
32393                     1          23.10  
38001                     1           1.45  
39132                     1           1.53  
40369                     1           2.69  
40387                     1           1.15  
41414                     1           1.60  
45857                     1           2.97  
49137                     1           1.04  
51816                     1           2.79  
53622                     1           1.06  
60241                     1           1.03  
71309                     1           1.20  
76655                     1           0.99  
79324                     1           2.42  
87981                     1           0.97  
88364                     1           1.40  
91510                     1           2.45  
93531                     1           1.45  
94873                     1           1.07  
95153                     1           0.99  
96566                     1           1.11

O parcelamento só é possível com cartão de crédito. Os demais tipos de pagamento têm somente 1 parcela.

In [40]:
order_payments[order_payments['payment_type'] != 'credit_card'].value_counts([order_payments['payment_type'], order_payments['payment_installments']])

payment_type  payment_installments
boleto        1                       19784
voucher       1                        5775
debit_card    1                        1529
not_defined   1                           3
dtype: int64

Nas compras com cartão de crédito, as compras à vista representam 33% do total, seguidas pelas compras parceladas em 2 vezes, correspondentes a 16%.

In [61]:
order_payments[order_payments['payment_type'] == 'credit_card'].value_counts(order_payments['payment_installments'], True)

payment_installments
1     0.331467
2     0.161638
3     0.136220
4     0.092428
10    0.069380
5     0.068221
8     0.055577
6     0.051045
7     0.021173
9     0.008386
12    0.001732
15    0.000964
18    0.000352
11    0.000299
24    0.000234
20    0.000221
13    0.000208
14    0.000195
17    0.000104
16    0.000065
21    0.000039
0     0.000026
22    0.000013
23    0.000013
dtype: float64

Quanto aos valores dos pagamentos, consolidamos o total por compra (order_id) e verificamos as principais medidas estatísticas.

In [71]:
order_total_value = order_payments[['order_id', 'payment_value']].groupby(['order_id']).sum()
order_total_value.describe()

payment_value
count   99440.000000
mean      160.990267
std       221.951257
min         0.000000
25%        62.010000
50%       105.290000
75%       176.970000
max     13664.080000

#### 1.2 Base de dados de produtos

In [73]:
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [74]:
products.tail()

product_id              product_category_name  \
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0

In [75]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


A base de dados possui 32951 registros e 9 características, sendo:
- 1 para identificação (product_id),
- 1 categórica (product_category_name),
- 3 quantitativas sobre a descrição de cada produto (product_name_lenght, product_description_lenght,product_photos_qty), e
- 4 quantitativas sobre peso e dimensões dos produtos (product_weight_g, product_length_cm, product_height_cm, product_width_cm)

Há 610 produtos com valores ausentes nas seguintes colunas:

In [86]:
products[products['product_category_name'].isna()].count()

product_id                    610
product_category_name           0
product_name_lenght             0
product_description_lenght      0
product_photos_qty              0
product_weight_g              609
product_length_cm             609
product_height_cm             609
product_width_cm              609
dtype: int64

Há 2 produtos sem informações numéricas de peso e dimensões.

In [93]:
products[products['product_weight_g'].isna()]

product_id product_category_name  \
8578   09ff539a621711667c43eba6a3bd8466                 bebes   
18851  5eb564652db742ff8f28759cd8d2652a                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
8578                  60.0                       865.0                 3.0   
18851                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
8578                NaN                NaN                NaN   
18851               NaN                NaN                NaN   

       product_width_cm  
8578                NaN  
18851               NaN

Dados estatísticos das características numéricas

In [76]:
products.describe()

product_name_lenght  product_description_lenght  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

Contagem de produtos por categoria

In [95]:
products_per_category = products['product_category_name'].value_counts()
products_per_category

cama_mesa_banho                  3029
esporte_lazer                    2867
moveis_decoracao                 2657
beleza_saude                     2444
utilidades_domesticas            2335
                                 ... 
fashion_roupa_infanto_juvenil       5
casa_conforto_2                     5
pc_gamer                            3
seguros_e_servicos                  2
cds_dvds_musicais                   1
Name: product_category_name, Length: 73, dtype: int64